File: LeetCode/696_计数二进制子串/696.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func countBinarySubstrings(s string) int {
	var num int
	var byteSlice = []byte(s)
	for i := 0; i < len(byteSlice) - 1; i++ {
		var sub []byte = byteSlice[i:]
		var flag bool = matchSubStrings(sub)
		if flag {
			num += 1
		}
	}
	return num
}

func matchSubStrings(sub []byte) bool {
	reg, _ := regexp.Compile("^(0+|1+)")
	sub_1 := reg.FindAllString(string(sub), -1)
	sub_2 := strings.Repeat(string(int(sub_1[0][0]) ^ 1), len(sub_1[0]))
	matchSub := sub_1[0] + sub_2
	if string(sub[0:len(matchSub)]) ==  matchSub {
		return true
	}
	return false
}

func main() {
	count := countBinarySubstrings("00110011")
	fmt.Println(count)
}

File: LeetCode/字符串/696_计数二进制子串/696.js
/**
 * @param {string} s
 * @return {number}
 */
var countBinarySubstrings = function(s) {
  let res = []

  // 匹配子串
  let match = (str) => {

    // 从头开始匹配 多个 0 或者 多个 1
    let j = str.match(/^(0+|1+)/)[0]
    // 再根据 多个 0 或者 多个 1 来创建符合条件的 子串
    // 例如 j = 000 ===> j$ = 111
    let j$ = (j[0] ^ 1).toString().repeat(j.length)

    // 正则匹配
    let reg = new RegExp(`^(${j}${j$})`)
    if (reg.test(str)) {
      return RegExp.$1
    } else {
      return ''
    }

    // let matchSub = j + j$
    // let subStr = str.slice(0, matchSub.length)
    // if (subStr === matchSub) {
    //   return matchSub
    // } else {
    //   return ''
    // }
    
  }

  for (let i = 0; i < s.length - 1; i++) {
    let sub = match(s.slice(i))
    if (sub) {
      res.push(sub)
    }
  }
  return res.length
};

let result = countBinarySubstrings("00110011")
console.log(result);